theseus/core/vectorizer.py

```diff
@@ -101,6 +102,16 @@ def __init__(self, objective: Objective):
             schema = _get_cost_function_schema(cost_fn)
             self._schema_dict[schema].append(wrapper)
 
+        # If all cost function groups have only one cost, there is no point in
+        # doing vectorization
+        if max([len(cfs) for cfs in self._schema_dict.values()]) <= 1:
```

Comment:
This is a good feature to have!

What's the behavior we expect for retraction in such a case?

I agree with Taosha, maybe in a future PR we can group only costs that repeat and leave others as is. Of course this will make the vectorization code even more complicated with maybe only a small advantage that the user doesn't need to do vectorize=False when setting up the optimizer.

Reply:
Good point about retraction, this would disable it unnecessarily, but I think this can be easily fixed. 

I actually think that Taosha's suggestion would not be too hard to implement. I'll give it a try for a couple of hours. If it gets too complicated we can add an issue for this.

Reply:
Changed, turned out to be even easier than I thought!